### PR TITLE
Build log subresource

### PR DESCRIPTION
**Pull request by csrwng**

### CODE REVIEW

File: pkg/build/api/types.go [hunk 1]
```diff
@@ -383,3 +383,12 @@ type BuildRequest struct {
...
 	Revision *SourceRevision `json:"revision,omitempty"`
 }
+
+// BuildLogOptions is the REST options for a build log
+type BuildLogOptions struct {
+	kapi.TypeMeta
+
+	// Follow if true indicates that the build log should be streamed until
+	// the build terminates.
+	Follow bool
```
Comment (mfojtik) [inline pkg/build/api/types.go]: `json:"follow,omitempty"`

Comment (smarterclayton) [inline pkg/build/api/types.go]: We don't add JSON tags to internal types.

Comment (ncdc) [inline pkg/build/api/types.go]: At least builds and images still have JSON on internal types - remove them?

Comment (smarterclayton) [inline pkg/build/api/types.go]: Sure, but that can be done in another pull.

Comment (ncdc) [inline pkg/build/api/types.go]: Agreed, just making sure that is needed.

**Comment by csrwng** [comment]
Made suggested changes

File: pkg/build/registry/buildlog/rest.go [hunk 2]
```diff
@@ -43,62 +48,97 @@ func (r RealPodControl) getPod(namespace, name string) (*kapi.Pod, error) {
...
+	}
+	build, err := r.BuildRegistry.GetBuild(ctx, name)
 	if err != nil {
-		return nil, nil, fielderrors.NewFieldNotFound("Build", id)
+		return nil, errors.NewNotFound("Build", name)
 	}
-
-	// TODO: these must be status errors, not field errors
-	// TODO: choose a more appropriate "try again later" status code, like 202
+	switch build.Status {
```
Comment (soltysh) [inline pkg/build/registry/buildlog/rest.go]: Will this switch handle a situation where build is being canceled when in New or Pending status?

File: pkg/build/registry/buildlog/rest.go [hunk 3]
```diff
@@ -43,62 +48,97 @@ func (r RealPodControl) getPod(namespace, name string) (*kapi.Pod, error) {
...
+	done := make(chan struct{})
+	errchan := make(chan error)
+	go func(ch <-chan watch.Event) {
+		for event := range ch {
+			obj, ok := event.Object.(*api.Build)
+			if !ok {
+				errchan <- fmt.Errorf("event object is not a build: %#v", event.Object)
+				break
+			}
+			switch obj.Status {
```
Comment (soltysh) [inline pkg/build/registry/buildlog/rest.go]: OK, my previous question (re canceled build) should be rather targeted at this switch.

File: pkg/cmd/server/origin/master.go [hunk 4]
```diff
@@ -211,6 +211,7 @@ func (c *MasterConfig) InstallProtectedAPI(container *restful.Container) []strin
 		"buildConfigs":             buildconfigregistry.NewREST(buildEtcd),
 		"buildConfigs/instantiate": buildConfigInstantiate,
 		"buildLogs":                buildlogregistry.NewREST(buildEtcd, c.BuildLogClient(), kubeletClient),
```
Comment (soltysh) [inline pkg/cmd/server/origin/master.go]: Aren't we going to deprecate this one or remove it completely?

File: pkg/build/registry/buildlog/rest.go (hunk 3 above)
Comment (soltysh) [inline pkg/build/registry/buildlog/rest.go]: Although, thinking about it more even in both of them.

File: pkg/cmd/server/origin/master.go (hunk 4 above)
Comment (csrwng) [inline pkg/cmd/server/origin/master.go]: We need to keep backwards compatibility

File: pkg/build/registry/buildlog/rest.go (hunk 3 above)
Comment (csrwng) [inline pkg/build/registry/buildlog/rest.go]: You're right. In both cases, I'd need to check for BuildStatusCancelled in addition to BuildStatusError. Thanks!

File: pkg/build/registry/buildlog/rest.go [hunk 5]
```diff
@@ -43,62 +48,97 @@ func (r RealPodControl) getPod(namespace, name string) (*kapi.Pod, error) {
...
+		Flush:       buildLogOpts.Follow,
+	}, nil
+}
 
-	// Pod in which build take place can't be in the Pending or Unknown phase,
-	// cause no containers are present in the Pod in those phases.
-	if pod.Status.Phase == kapi.PodPending || pod.Status.Phase == kapi.PodUnknown {
-		return nil, nil, fielderrors.NewFieldInvalid("Pod.Status", pod.Status.Phase, "must be Running, Succeeded or Failed")
+func (r *REST) waitForBuild(ctx kapi.Context, build *api.Build) error {
+	fieldSelector := fields.Set{"name": build.Name}.AsSelector()
```
Comment (smarterclayton) [inline pkg/build/registry/buildlog/rest.go]: Should be "metadata.name"

**Comment by smarterclayton** [comment]
Are the existing tests sufficient to prove this still works?

File: pkg/cmd/server/origin/master.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: You can remove it from v1beta3

File: pkg/build/registry/buildlog/rest.go [hunk 6]
```diff
@@ -43,62 +48,97 @@ func (r RealPodControl) getPod(namespace, name string) (*kapi.Pod, error) {
...
+		return nil, errors.NewNotFound("Build", name)
 	}
-
-	// TODO: these must be status errors, not field errors
-	// TODO: choose a more appropriate "try again later" status code, like 202
+	switch build.Status {
+	// Build has not launched, wait til it runs
+	case api.BuildStatusNew, api.BuildStatusPending:
+		if !buildLogOpts.Wait {
+			return nil, nil
```
Comment (smarterclayton) [inline pkg/build/registry/buildlog/rest.go]: Should be returning a 204, not a 200.  Add a TODO

File: pkg/build/registry/buildlog/rest.go [hunk 7]
```diff
@@ -43,62 +48,97 @@ func (r RealPodControl) getPod(namespace, name string) (*kapi.Pod, error) {
...
+			}
+		}
+	}(w.ResultChan())
+	select {
+	case err := <-errchan:
+		return err
+	case <-done:
+		return nil
+	case <-time.After(r.Timeout):
+		return fmt.Errorf("Timed out waiting for build.")
```
Comment (smarterclayton) [inline pkg/build/registry/buildlog/rest.go]: "timed out waiting for build"

`error` should not be sentences

File: pkg/build/registry/buildlog/rest.go (hunk 6 above)
Comment (csrwng) [inline pkg/build/registry/buildlog/rest.go]: We actually return a 204 if the ResourceStreamer returns a nil stream (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/apiserver/apiserver.go#L250). Will just change how we return this.

Comment (smarterclayton) [inline pkg/build/registry/buildlog/rest.go]: K. Just checking

**Comment by csrwng** [comment]
Added additional tests to the REST resource

**Comment by csrwng** [comment]
Updated with v1beta3 and commands.
@smarterclayton should be ready

**Comment by soltysh** [comment]
LGTM.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**